torch/csrc/jit/frontend/source_range.h

```diff
@@ -105,15 +113,41 @@ struct Source {
   std::shared_ptr<SourceRangeUnpickler> gen_ranges_;
 };
 
-// A SourceRange is a view into a Source, that points to a subset of the source,
-// specified by `start` and `end` byte offsets into the source text.
+struct SourceView : public Source {
```

Comment:
I don't think this is a particularly good use of inheritance. A `Source` owns its string whereas a `SourceView` does not; handing a `SourceView` to code that expects a `Source` could easily end in tears. CC @ezyang  for the Liskov Substitution Principle lecture :)

Reply:
Yeah, that's a real possibility but I think it should be fine in this case because the derived class already clearly states that it is a view and does not own the text. Thus caller is responsible for making the lifetime correct. Let me know what you think.

Reply:
> The derived class already clearly states that it is a view

Code written to expect a `Source` knows nothing about `SourceView` and thus can't see those statements. It may, for example, try to retain a `Source` for later use. Inheritance should not be used merely for implementation reuse.

Reply:
Sorry for the delay. I have been on PTO recently.

What if we refactor code to make Source and SourceView both inherit from SourceBase, which makes no guarantee about ownership?

Reply:
Sounds like a good plan, making sure we are all on the same page:

1) `SourceView` inherits from `Source`, overriding its `text` method, allowing implicit construction of `SourceView` from `Source`
2) All current clients of `Source` should use `SourceView`, except for `ErrorReport`, which should always retain a real copy of source text.
3) Producers of `Source` stay unchanged.

@swolchok @suo sounds good?

Reply:
> SourceView inherits from Source, overriding its text method, allowing implicit construction of SourceView from Source

Inheriting like that won't allow implicit construction of SourceView from Source, it would allow implicit construction of Source from SourceView.